subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java

```diff
@@ -238,8 +238,8 @@ private void validatePluginDeclarations(Project project, GradlePluginDevelopment
         });
     }
 
-    private void configurePluginValidations(Project project) {
-        SourceSet mainSourceSet = project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets().getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+    private void configurePluginValidations(Project project, GradlePluginDevelopmentExtension extension) {
+        SourceSet pluginSourceSet = extension.getPluginSourceSet();
```

Comment:
Shouldn't this lookup be lazy? If we do it here, then changing the `pluginSourceSet` on the extension won't have any effect, right?

Let's also add a test for this new behaviour.

Reply:
Well spotted. Added test and fix.